feat(types): add declarations for `ModuleContainer` utils

kawahara opened the pull request:
## Types of changes
- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)


## Description
Resolve #9163 


## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

kawahara committed:
feat(types): add type declaration for ModuleContainer

pi0 commented on packages/types/config/module.d.ts:
```diff
@@ -17,10 +19,35 @@ interface ExtendFunctionContext {
-type ExtendFunction = (config: WebpackConfiguration, ctx: ExtendFunctionContext) => void
+export interface ModuleTemplateConfig {
+  src: string,
+  fileName?: string,
```
- We have both `fileName?` and `filename?`
- Options can be types as `Record<string, any>`
- We also allow `[key: string]: any` additional properties for custom purposes

pi0 commented on packages/types/config/module.d.ts:
```diff
@@ -17,10 +19,35 @@ interface ExtendFunctionContext {
 interface ModuleThis {
-  extendBuild(fn: ExtendFunction): void
+  addTemplate(template: ModuleTemplate): ModuleTemplateDest
+  addPlugin(template: ModuleTemplate): void
```
We need an interface extending `ModuleTemplate` for plugins with `ssr?: boolean` and `mode?: 'all' | 'client' | 'server'`

kawahara committed:
feat(types): fix config attributes for template

- add `filename` that is alias for `fileName` to `ModuleTemplateConfig`
- change types of ModuleTemplateConfig.options to Record
- allow additional keys for `ModuleTemplateConfig`
- Make interface for plugin template. it has ssr and mode field

pi0 commented:
Thanks!
